### PR TITLE
new event info type email

### PR DESCRIPTION
**Pull request by edevil**
Added new event info type for email events. This will contain the from, to and raw size of the email.

### CODE REVIEW

File: src/workerd/io/trace.h [hunk 1]
```diff
@@ -112,6 +112,18 @@ class Trace final : public kj::Refcounted {
     void copyTo(rpc::Trace::QueueEventInfo::Builder builder);
   };
 
+class EmailEventInfo {
```
Comment (harrishancock) [inline src/workerd/io/trace.h:115]: Nit: incorrect indentation

Comment (edevil) [reply src/workerd/io/trace.h]: Done.

File: src/workerd/api/trace.h [hunk 2]
```diff
@@ -209,6 +211,25 @@ class TraceItem::QueueEventInfo final: public jsg::Object {
...
+  explicit EmailEventInfo(kj::Own<Trace> trace, const Trace::EmailEventInfo& eventInfo);
+
+  kj::StringPtr getMailFrom();
+  kj::StringPtr getRcptTo();
+  uint32_t getRawSize();
+
+  JSG_RESOURCE_TYPE(EmailEventInfo) {
+    JSG_READONLY_INSTANCE_PROPERTY(mailFrom, getMailFrom);
+    JSG_READONLY_INSTANCE_PROPERTY(rcptTo, getRcptTo);
+    JSG_READONLY_INSTANCE_PROPERTY(rawSize, getRawSize);
```
Comment (jasnell) [inline src/workerd/api/trace.h:225]: Let's make these `JSG_READONLY_PROTOTYPE_PROPERTY`. Recent changes to garbage collection handling make instance properties significantly less desirable.

Comment (edevil) [reply src/workerd/api/trace.h]: Done.

**Comment by harrishancock** [comment]
@jasnell, unrelated to this specific PR, but should the prototype property change (https://github.com/cloudflare/workerd/pull/249#discussion_r1062634607) be extended to all the other similar EventInfo types, potentially with a compatibility flag?

**Review by harrishancock — APPROVED** [review]

**Comment by jasnell** [comment]
> ... but should the prototype property change (https://github.com/cloudflare/workerd/pull/249#discussion_r1062634607) be extended to all the other similar EventInfo types, potentially with a compatibility flag?

We should likely start transitioning those, yes, but not sure how much of a priority it is.

File: src/workerd/api/trace.h [hunk 3]
```diff
@@ -209,6 +211,25 @@ class TraceItem::QueueEventInfo final: public jsg::Object {
...
+
+  JSG_RESOURCE_TYPE(EmailEventInfo) {
+    JSG_READONLY_PROTOTYPE_PROPERTY(mailFrom, getMailFrom);
+    JSG_READONLY_PROTOTYPE_PROPERTY(rcptTo, getRcptTo);
+    JSG_READONLY_PROTOTYPE_PROPERTY(rawSize, getRawSize);
+  }
+
+private:
+  kj::Own<Trace> trace;
+  const Trace::EmailEventInfo& eventInfo;
```
Comment (jasnell) [inline src/workerd/api/trace.h:230]: I don't believe this is a new problem at all but I do generally have a concern about using const refs in jsg::Objects like this. What guarantee do we have that the ref is going to stay alive for as long as the JS wrapper? For instance, if someone happens to store the info object at global scope, for instance, what happens when the properties are accessed?

Comment (harrishancock) [reply src/workerd/api/trace.h:230]: I agree, we should look at this.

**Pull request merged**